src/ui/components/login/login_form.rs

```diff
@@ -1,85 +1,213 @@
 
-  let QueryResult { refetch, .. } = use_site_state();
+  let login = create_server_action::<LoginFn>();
 
-  create_isomorphic_effect(move |_| {
-    if login_is_success.get() {
-      refetch();
+  let username_validation = create_rw_signal::<String>("".into());
+  let password_validation = create_rw_signal::<String>("".into());
 
-      cfg_if! {
-        if #[cfg(feature = "ssr")] {
-          leptos_actix::redirect("/");
-        } else {
-          let navigate = leptos_router::use_navigate();
+  let ssr_error = move || query.with(|params| params.get("error").cloned());
 
-          navigate("/", leptos_router::NavigateOptions { replace: true, ..Default::default() })
+  if let Some(e) = ssr_error() {
+    let le = serde_json::from_str::<LemmyAppError>(&e[..]);
+
+    match le {
+      Ok(e) => match e {
+        LemmyAppError {
+          error_type: LemmyAppErrorType::EmptyUsername,
+          ..
+        } => {
+          username_validation.set("input-error".to_string());
         }
-      }
+        LemmyAppError {
+          error_type: LemmyAppErrorType::EmptyPassword,
+          ..
+        } => {
+          password_validation.set("input-error".to_string());
+        }
+        _ => {}
+      },
+      Err(_) => {}
     }
-  });
+  }
+
+  let on_submit = move |ev: SubmitEvent| {
+    ev.prevent_default();
+    error.set(None);
+
+    create_local_resource(
+      move || (name.get(), password.get()),
+      move |(name, password)| async move {
+        let req = Login {
+          username_or_email: name.into(),
+          password: password.into(),
+          totp_2fa_token: None,
+        };
+        let result = try_login(req.clone()).await;
+        match result {
+          Ok(LoginResponse { jwt: Some(jwt), .. }) => {
+            let _ = set_cookie(
+              "jwt",
+              &jwt.clone().into_inner(),
+              &core::time::Duration::from_secs(604800),
+            )
+            .await;
+            user.set(Some(true));
+            leptos_router::use_navigate()("/", Default::default());
+          }
+          Ok(LoginResponse { jwt: None, .. }) => {
+            error.set(Some(LemmyAppError {
+              error_type: LemmyAppErrorType::MissingToken,
+              content: String::default(),
+            }));
+          }
+          Err(e) => {
+            error.set(Some(e.clone()));
+            password_validation.set("".to_string());
+            username_validation.set("".to_string());
+
+            match e {
+              LemmyAppError {
+                error_type: LemmyAppErrorType::EmptyUsername,
+                ..
+              } => {
+                username_validation.set("input-error".to_string());
+              }
+              LemmyAppError {
+                error_type: LemmyAppErrorType::EmptyPassword,
+                ..
+              } => {
+                password_validation.set("input-error".to_string());
+              }
+              _ => {}
+            }
```

Comment:
The following accomplishes the same thing while (in my opinion at least) being easier to read:

```suggestion
            match e.error_type {
              LemmyAppErrorType::EmptyUsername => {
                username_validation.set("input-error".to_string());
              }
              LemmyAppErrorType::EmptyPassword => {
                password_validation.set("input-error".to_string());
              }
              _ => {}
            }
```

You may also like the following:
```suggestion
            if e.error_type == LemmyAppErrorType::EmptyUsername {
                username_validation.set("input-error".to_string());
            } else if e.error_type == LemmyAppErrorType::EmptyPassword {
                password_validation.set("input-error".to_string());
            }
```

The latter assumes `LemmyAppErrorType` implements `PartialEq` and `Eq`.

Reply:
i should have marked the PR as draft again before pushing update - this code may have been removed or changed